dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala

```diff
@@ -0,0 +1,132 @@
+
+   def discardSize(): Int = _discardSize
+
+   def word2Index(): Map[String, Int] = _word2index
+
+   def index2Word(): Map[Int, String] = _index2word
+
+   def vocabulary(): Array[String] = _vocabulary.toArray
+
+   def discardVocab(): Array[String] = _disvardVocab.toArray
+
+   def getIndex(word: String): Int = {
+     _word2index.getOrElse(word, _vocabSize)
+   }
+
+   def getWord(index: Float): String = {
+     getWord(index.toInt)
+   }
+
+   def getWord(index: Double): String = {
+     getWord(index.toInt)
+   }
+
+   def getWord(index: Int): String = {
+     _index2word.getOrElse(index,
+       _disvardVocab(Random.nextInt(_discardSize)))
+   }
+
+   def print(): Unit = {
+     _word2index.foreach(x =>
+       logger.info(x._1 + " -> " + x._2))
+   }
+
+   def printDiscard(): Unit = {
+     _disvardVocab.foreach(x =>
+       logger.info(x))
+   }
+
+   def this(sentences: Array[Array[String]],
+            vocabSize: Int) = {
+     this()
+     val freqDict = sentences
+       .flatMap(x => x)
+       .foldLeft(Map.empty[String, Int]) {
+         (count, word) => count + (word -> (count.getOrElse(word, 0) + 1))
+       }.toSeq.sortBy(_._2)
+
+     // Select most common words
+     val length = math.min(vocabSize, freqDict.length)
+     _vocabulary = freqDict.drop(freqDict.length - length).map(_._1)
+     _vocabSize = _vocabulary.length
+     _word2index = _vocabulary.zipWithIndex.toMap
+     _index2word = _word2index.map(x => (x._2, x._1))
+     _disvardVocab = freqDict.take(freqDict.length - length).map(_._1)
+     _discardSize = _disvardVocab.length
+   }
+
+   def this(directory: String) = {
+     this()
+
+     val dictionaryFile = new File(directory, "dictionary.txt")
+     if (!dictionaryFile.exists()) {
+       throw new IllegalArgumentException("dictionary file not exists!")
+     }
+     val discardFile = new File(directory, "discard.txt")
+     if (!discardFile.exists()) {
+       throw new IllegalArgumentException("discard file not exists!")
+     }
+
+     import scala.io.Source
+     _word2index = Source.fromFile(dictionaryFile.getAbsolutePath)
+       .getLines.map(_.stripLineEnd.split("->", -1))
+       .map(fields => fields(0).stripSuffix(" ") -> fields(1).stripPrefix(" ").toInt)
+       .toMap[String, Int]
+     _index2word = _word2index.map(x => (x._2, x._1))
+     _vocabulary = _word2index.keys.toSeq
+     _vocabSize = _word2index.size
+     _disvardVocab = Source.fromFile(discardFile.getAbsolutePath)
+       .getLines().toSeq
+     _discardSize = _disvardVocab.length
+   }
+
+   val logger = Logger.getLogger(getClass)
+   private var _vocabSize: Int = 0
+   private var _discardSize: Int = 0
+   private var _word2index: Map[String, Int] = null
+   private var _index2word: Map[Int, String] = null
+   private var _vocabulary: Seq[String] = null
+   private var _disvardVocab: Seq[String] = null
+}
+
+object Dictionary {
+  def apply(sentences: Array[Array[String]],
+            vocabSize: Int)
```

Comment:
Is it better to add arguments such as 'needDiscard: Boolean = true' to set whether the user need discardVocab? I'm not sure :-)